src/plugins/chatMarkers/ChatMarkersPlugin.ts

```diff
@@ -0,0 +1,410 @@
+               }
+            }
+         }
+      }
+      return true;
+   }
+
+   // mark messages, depending on the chat marker we get, according to XEP-0333
+   private onChatMarkersMessage = (stanza: string) => {
+      let element = $(stanza).find(RECEIVED + Namespace.getFilter(CHATMARKERS));
+      let received = false;
+      let displayed = false;
+      let acknowledged = false;
+      if (element.length !== 0) {
+         this.pluginAPI.Log.debug(`message has "${RECEIVED}" element. Yaay! =)`);
+         received = true;
+      } else {
+         element = $(stanza).find(DISPLAYED + Namespace.getFilter(CHATMARKERS));
+         if (element.length !== 0) {
+            this.pluginAPI.Log.debug(`message has "${DISPLAYED}" element. Yaay! =)`);
+            displayed = true;
+            received = true;
+         } else {
+            element = $(stanza).find(ACKNOWLEDGED + Namespace.getFilter(CHATMARKERS));
+            if (element.length !== 0) {
+               this.pluginAPI.Log.debug(`message has "${ACKNOWLEDGED}" element. Yaay! =)`);
+               acknowledged = true;
+               displayed = true;
+               received = true;
+            }
+         }
+      }
+      let msgId = element.attr(ID);
+      if (!!msgId) {
+         this.pluginAPI.Log.debug(`message has "${ID}" attribute (${ID}: ${msgId}). Yaay! =)`);
+         let from = $(stanza).attr(FROM);
+         if (!!from) {
+            this.pluginAPI.Log.debug(`message has "${FROM}" attribute (${FROM}: ${from}). Yaay! =)`);
+            if ($(stanza).attr('type') !== Message.MSGTYPE.GROUPCHAT) {
+               let contact = this.pluginAPI.getContact(new JID(from));
+               if (contact !== undefined) {
+                  let transcript = contact.getChatWindow().getTranscript();
+                  let msg = transcript.getMessage(msgId);
+                  while (!!msg) {
+                     if (msg.getDirection() === DIRECTION.OUT) {
+                        if (received) {
+                           msg.chatMarkersReceived();
+                        }
+                        if (displayed) {
+                           msg.chatMarkersDisplayed();
+                        }
+                        if (acknowledged) {
+                           msg.chatMarkersAcknowledged();
+                        }
+                     }
+                     try {
+                        msg = transcript.getMessage(msg.getNextId());
+                     } catch (error) {
+                        break;
+                     }
+                  }
+               }
+            }
+         }
+      }
+      return true;
+   }
+
+   private chatWindowInitializedHook(chatWindow?: ChatWindow, contact?: Contact) {
+      let transcript = chatWindow.getTranscript();
+      let msg = transcript.getFirstMessage();
+      while (!!msg) {
+         let chatWindowMsg = chatWindow.getChatWindowMessage(msg) as ChatWindowMessage;
```

Comment:
Why is the cast needed?

Reply:
For type safety, as ChatWindow.getChatWindowMessage() doesn't declare a return type. But as this method returns a ChatWindowMessage for sure, it's probably not necessary to cast.

Reply:
The better solution would be to change the signature of `ChatWindow.getChatWindowMessage()` :wink: